### PR TITLE
Merge Keras model import into main branch

### PR DESCRIPTION
**Pull request by turambar**
Keras model import is ready to be included in the DL4J release. At the very least, it passes obvious unit tests covering import of basic layers (dense, convolutional, LSTM) and constructing Sequential models. Functionality is segregated in the modelimport module so does not impact any other DL4J functionality. Still lots of testing to do and functionality to add.

### CODE REVIEW

File: deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras/ModelConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        String configFilename = classLoader.getResource("keras/config/cnn_tf_config.json").getPath();
+        importKerasConfigTest(configFilename);
+    }
+
+    @Test
+    public void importKerasConvnetTheanoConfigTest() throws Exception {
+        ClassLoader classLoader = getClass().getClassLoader();
+        String configFilename = classLoader.getResource("keras/config/cnn_th_config.json").getPath();
+        importKerasConfigTest(configFilename);
+    }
```
Comment (agibsonccc) [inline deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras/ModelConfigurationTest.java]: Use ClassPathResource here. Way more robust.

File: deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+                break;
+            case "Convolution2D":
+                layer = buildConvolutionLayer(kerasConfig);
+                break;
+            case "MaxPooling2D":
+                layer = buildSubsamplingLayer(kerasConfig);
+                break;
+            case "Flatten":
+                log.warn("DL4J adds reshaping layers during model compilation");
+                break;
```
Comment (agibsonccc) [inline deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java]: May want to link to something here?

File: deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+     *
+     * @param kerasConfig      Map containing Keras Dense layer properties
+     * @return                 DL4J DenseLayer configuration
+     * @throws NotImplementedException
+     * @see DenseLayer
+     */
+    public static DenseLayer buildDenseLayer(Map<String,Object> kerasConfig)
+        throws NotImplementedException {
+        DenseLayer.Builder builder = new DenseLayer.Builder()
+                .nOut((int)kerasConfig.get("output_dim"));
```
Comment (agibsonccc) [inline deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java]: These "magic strings" should be constants with public final static SOME_VARIABLE_NAME.

File: deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras/ModelConfigurationTest.java (hunk 1 above)
Comment (turambar) [inline deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras/ModelConfigurationTest.java]: Done. Change pushed.

**Comment by AlexDBlack** [comment]
Notes from review:
- pom.xml
    - junit must be test scope
    - logback-classic should have version
- Javadoc on the IncompatibleKerasConfigurationException
    - People will hit this, and the javadoc should help answer the questions "why?" and "what should I do now?"
- Weight init, Keras -> DL4J mapping:
    - glorot_normal is xavier
    - glorot_uniform is xavier_uniform
    - he_normal is relu
    - he_uniform is relu_uniform
    - Normal and lecun uniform *might* map uniform and sigmoid uniform? check math on that
- String constants should ideally be public/private static final fields in LayerConfig
- Is Keras LSTM the version with or without the peephole connections? Ours has them, and we should be able to import both (if keras version is without the peephole connections: ours should be set to 0.0)
- CNN border mode: we support 'valid' only, not 'same' (currently). I don't see this checked anywhere
- Test resources should be placed here: https://github.com/deeplearning4j/dl4j-test-resources/commits/master
- Re: tests - I'd be more comfortable if we had a (simple) full test of a trained model. Say something like trained mnist model -> import to dl4j. Then check:
    - Same output/predictions for a given same input
    - Same score (loss function value) given same input
    - Ideally: also check same weight gradients for both (first 1 is most important)
    - Eventually we'll want the same for dense and LSTM layers too...
- How are we handling updaters? I assume it's not part of the model itself. Presumably we are using DL4J's default?
    - The behaviour here should at least be documented somewhere

File: deeplearning4j-modelimport/pom.xml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+            <version>${slf4j.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>${junit.version}</version>
+        </dependency>
```
Comment (AlexDBlack) [inline deeplearning4j-modelimport/pom.xml]: Junit must have <scope>test</scope>

File: deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/IncompatibleKerasConfigurationException.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package org.deeplearning4j.nn.modelimport.keras;
+
+
+public class IncompatibleKerasConfigurationException extends RuntimeException {
```
Comment (AlexDBlack) [inline deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/IncompatibleKerasConfigurationException.java]: Javadoc: People will hit this exception, and the javadoc should help answer the questions "why?" and "what should I do now?"
Maybe mention that it's WIP and to visit the dl4j issues page or gitter?

File: deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+                    break;
+                case "lecun_uniform":
+                case "normal":
+                case "identity":
+                case "orthogonal":
+                case "glorot_normal":
+                case "glorot_uniform":
+                case "he_normal":
+                case "he_uniform":
+                default:
```
Comment (AlexDBlack) [inline deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java]: Weight init, Keras -> DL4J mapping:
glorot_normal is xavier
glorot_uniform is xavier_uniform
he_normal is relu
he_uniform is relu_uniform
Normal and lecun uniform might map uniform and sigmoid uniform? check math on that

**Comment by saudet** [comment]
I've released stuff on the Maven Central Repository, so the following should work now:
```xml
    <properties>
        <hdf5.version>1.10.0-patch1</hdf5.version>
        <javacpp-presets.version>1.2</javacpp-presets.version>
        <javacpp.version>1.2.5</javacpp.version>
    </properties>
    <dependencies>
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>javacpp</artifactId>
            <version>${javacpp.version}</version>
        </dependency>
        <dependency>
            <groupId>org.bytedeco.javacpp-presets</groupId>
            <artifactId>hdf5</artifactId>
            <version>${hdf5.version}-${javacpp-presets.version}</version>
        </dependency>
        <dependency>
            <groupId>org.bytedeco.javacpp-presets</groupId>
            <artifactId>hdf5</artifactId>
            <version>${hdf5.version}-${javacpp-presets.version}</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>org.bytedeco.javacpp-presets</groupId>
            <artifactId>hdf5</artifactId>
            <version>${hdf5.version}-${javacpp-presets.version}</version>
            <classifier>macosx-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>org.bytedeco.javacpp-presets</groupId>
            <artifactId>hdf5</artifactId>
            <version>${hdf5.version}-${javacpp-presets.version}</version>
            <classifier>windows-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>org.bytedeco.javacpp-presets</groupId>
            <artifactId>hdf5</artifactId>
            <version>${hdf5.version}-${javacpp-presets.version}</version>
            <classifier>linux-ppc64le</classifier>
        </dependency>
    </dependencies>
```
All that is basically going to move to the platform artifacts in the next major release of JavaCPP later this month, but the effect is pretty much the same. Just a bit more verbose. Let me know if this doesn't work for some reason! Thanks

File: deeplearning4j-modelimport/pom.xml (hunk 4 above)
Comment (turambar) [inline deeplearning4j-modelimport/pom.xml]: Fixed.

File: deeplearning4j-modelimport/pom.xml [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            <artifactId>nd4j-native</artifactId>
+            <version>${nd4j.version}</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (turambar) [inline deeplearning4j-modelimport/pom.xml]: Fixed.

File: deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/IncompatibleKerasConfigurationException.java (hunk 5 above)
Comment (turambar) [inline deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/IncompatibleKerasConfigurationException.java]: Punt until next release

File: deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java (hunk 2 above)
Comment (turambar) [inline deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java]: @agibsonccc Adding a link to github repos location showing code where we add preprocessing layers: https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/MultiLayerConfiguration.java#L429

Does that suffice?

File: deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java (hunk 3 above)
Comment (turambar) [inline deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java]: No more magic strings. Still some magic constants. They're probably OK as they're things like 0.0 or 1.0

File: deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/IncompatibleKerasConfigurationException.java (hunk 5 above)
Comment (turambar) [inline deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/IncompatibleKerasConfigurationException.java]: Added Keras model import doc page with pithy explanation. Added link to that in javadoc comments and error message.

**Comment by turambar** [comment]
Merged after making above changes and acquiring @AlexDBlack's say so.
